File: 05-llm-api/examples.py
"""
Минимальный клиент GigaChat без сторонних SDK — на голом requests.

Нужно:
  pip install requests
  export GIGACHAT_AUTH_KEY="<base64(client_id:client_secret)>"
  # ключ выдают в личном кабинете SberDev

Запуск: python examples.py
"""

import os
import time
import uuid
import requests


AUTH_URL = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth"
CHAT_URL = "https://gigachat.devices.sberbank.ru/api/v1/chat/completions"


class GigaChatClient:
    def __init__(self, auth_key: str, scope: str = "GIGACHAT_API_PERS") -> None:
        self.auth_key = auth_key
        self.scope = scope
        self._token: str | None = None
        self._token_expires_at: float = 0.0

    def _get_token(self) -> str:
        # Освежаем за минуту до истечения, чтобы не нарваться на 401
        if self._token and time.time() < self._token_expires_at - 60:
            return self._token

        resp = requests.post(
            AUTH_URL,
            headers={
                "Authorization": f"Basic {self.auth_key}",
                "RqUID": str(uuid.uuid4()),
                "Content-Type": "application/x-www-form-urlencoded",
            },
            data={"scope": self.scope},
            verify=False,  # для прода — путь к сертификату Минцифры
            timeout=10,
        )
        resp.raise_for_status()
        data = resp.json()
        self._token = data["access_token"]
        # expires_at в миллисекундах
        self._token_expires_at = data["expires_at"] / 1000
        return self._token

    def chat(
        self,
        messages: list[dict],
        model: str = "GigaChat",
        temperature: float = 0.7,
        max_tokens: int = 1024,
    ) -> str:
        token = self._get_token()
        resp = requests.post(
            CHAT_URL,
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
            },
            json={
                "model": model,
                "messages": messages,
                "temperature": temperature,
                "max_tokens": max_tokens,
            },
            verify=False,
            timeout=60,
        )
        resp.raise_for_status()
        data = resp.json()
        return data["choices"][0]["message"]["content"]


def main() -> None:
    auth_key = os.environ.get("GIGACHAT_AUTH_KEY")
    if not auth_key:
        print("Установи переменную окружения GIGACHAT_AUTH_KEY")
        return

    client = GigaChatClient(auth_key)
    messages = [
        {"role": "system", "content": "Ты лаконичный помощник, отвечай в 1-2 предложения."},
        {"role": "user", "content": "Что такое JSON?"},
    ]
    answer = client.chat(messages)
    print("Ответ:", answer)


if __name__ == "__main__":
    # отключим warning о verify=False для примера
    import urllib3

    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
    main()

File: 05-llm-api/examples.go
// Минимальный клиент GigaChat на голом net/http.
// Запуск:
//   export GIGACHAT_AUTH_KEY="<base64(client_id:client_secret)>"
//   go run examples.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/google/uuid"
)

const (
	authURL = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth"
	chatURL = "https://gigachat.devices.sberbank.ru/api/v1/chat/completions"
)

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type chatRequest struct {
	Model       string    `json:"model"`
	Messages    []Message `json:"messages"`
	Temperature float64   `json:"temperature"`
	MaxTokens   int       `json:"max_tokens"`
}

type chatResponse struct {
	Choices []struct {
		Message Message `json:"message"`
	} `json:"choices"`
}

type tokenResponse struct {
	AccessToken string `json:"access_token"`
	ExpiresAt   int64  `json:"expires_at"` // миллисекунды
}

type GigaChatClient struct {
	authKey   string
	scope     string
	http      *http.Client
	token     string
	expiresAt time.Time
}

func NewGigaChatClient(authKey string) *GigaChatClient {
	// Для прода: указать корневой сертификат Минцифры через RootCAs.
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	return &GigaChatClient{
		authKey: authKey,
		scope:   "GIGACHAT_API_PERS",
		http:    &http.Client{Timeout: 60 * time.Second, Transport: tr},
	}
}

func (c *GigaChatClient) getToken() (string, error) {
	if c.token != "" && time.Now().Before(c.expiresAt.Add(-time.Minute)) {
		return c.token, nil
	}

	form := url.Values{"scope": {c.scope}}
	req, _ := http.NewRequest("POST", authURL, strings.NewReader(form.Encode()))
	req.Header.Set("Authorization", "Basic "+c.authKey)
	req.Header.Set("RqUID", uuid.NewString())
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := c.http.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("auth %d: %s", resp.StatusCode, body)
	}

	var tr tokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&tr); err != nil {
		return "", err
	}
	c.token = tr.AccessToken
	c.expiresAt = time.UnixMilli(tr.ExpiresAt)
	return c.token, nil
}

func (c *GigaChatClient) Chat(messages []Message) (string, error) {
	token, err := c.getToken()
	if err != nil {
		return "", err
	}

	body, _ := json.Marshal(chatRequest{
		Model:       "GigaChat",
		Messages:    messages,
		Temperature: 0.7,
		MaxTokens:   1024,
	})
	req, _ := http.NewRequest("POST", chatURL, bytes.NewReader(body))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.http.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		b, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("chat %d: %s", resp.StatusCode, b)
	}

	var cr chatResponse
	if err := json.NewDecoder(resp.Body).Decode(&cr); err != nil {
		return "", err
	}
	if len(cr.Choices) == 0 {
		return "", fmt.Errorf("пустой ответ")
	}
	return cr.Choices[0].Message.Content, nil
}

func main() {
	authKey := os.Getenv("GIGACHAT_AUTH_KEY")
	if authKey == "" {
		fmt.Println("Установи GIGACHAT_AUTH_KEY")
		return
	}

	client := NewGigaChatClient(authKey)
	answer, err := client.Chat([]Message{
		{Role: "system", Content: "Ты лаконичный помощник, отвечай в 1-2 предложения."},
		{Role: "user", Content: "Что такое JSON?"},
	})
	if err != nil {
		fmt.Println("ошибка:", err)
		return
	}
	fmt.Println("Ответ:", answer)
}
